corehq/util/validation.py

```diff
@@ -19,3 +22,36 @@ def is_url_or_host_banned(url_or_host):
     # use this check to help site admins ensure they're not making any obvious
     # mistakes.
     return any([re.search(regex, url_or_host) for regex in BANNED_HOST_REGEX])
+
+
+@deconstructible
+class JSONSchemaValidator:
+    """Field level validation for JSONField against a JSON Schema"""
+
+    def __init__(self, schema):
+        self.schema = schema
+        self.schema_validator_class = jsonschema.validators.validator_for(schema)
+        self.schema_validator_class.check_schema(schema)
+
+    def __call__(self, value):
+        errors = self.schema_validator_class(self.schema).iter_errors(value)
+
+        def _extract_errors(_errors):
+            for error in _errors:
+                if error.context:
+                    return self._extract_errors(error.context)
+
+                message = str(error).replace("\n\n", ": ").replace("\n", "")
+                django_errors.append(ValidationError(message))
+
+        django_errors = []
+        _extract_errors(errors)
+        if django_errors:
+            raise ValidationError(django_errors)
+
+        return value
+
+    def __eq__(self, other):
```

Comment:
> To prevent a new migration from being created each time makemigrations is run, you should also add a __eq__() method to the decorated class. This function will be called by Django’s migration framework to detect changes between states.

Was wondering what this was for.  Makes sense now